### PR TITLE
Add test to check all or none implement default/validate helpers

### LINKED ISSUE #3986: Detection of missing hand-written code
**Describe the current behavior**

Some of our extensions need to be implemented on every version of a resource, or none at all.

For example, the `genruntime.Defaulter` interface was implemented for API version `2021-05-15` of `documentdb.SqlRoleAssignment` and therefore should also have been implemented for API version `2023-11-15` when that was imported.

Failing to implement that lead to some hairy errors that required considerable debugging.

**Describe the improvement**

I'm wondering if we can write a _convention-test_ that checks all our resources to verify that if one version implements  `genruntime.Defaulter`, all versions do? Can we generalize this to other interface implementations?

### PR DESCRIPTION
**Pull request by matthchr**
Fixes #3986.

- [ ] this PR contains documentation
- [x] this PR contains tests
- [ ] this PR contains YAML Samples

### CODE REVIEW

**Comment by matthchr** [comment]
I tested by locally updating types to violate the check and ensure that both tests failed correctly if a mistake was made.

File: v2/internal/tests/api_linting_test.go [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+	}
+
+	// Check if the type has the three required methods:
+	// CreateValidations() []func(ctx context.Context, obj T) (admission.Warnings, error)
+	// UpdateValidations() []func(ctx context.Context, oldObj T, newObj T) (admission.Warnings, error)
+	// DeleteValidations() []func(ctx context.Context, obj T) (admission.Warnings, error)
+
+	createMethod, hasCreate := objType.MethodByName("CreateValidations")
+	updateMethod, hasUpdate := objType.MethodByName("UpdateValidations")
+	deleteMethod, hasDelete := objType.MethodByName("DeleteValidations")
```
Comment (theunrepentantgeek) [inline v2/internal/tests/api_linting_test.go]: This method is full of repeated triples - a code smell indicating there's potentially a common helper function or two to extract.

Comment (matthchr) [reply v2/internal/tests/api_linting_test.go]: Good callout - fixed.

**Review by theunrepentantgeek — APPROVED** [review]
